pulp_ansible/app/galaxy/v3/views.py

```diff
@@ -126,68 +157,92 @@ class CollectionViewSet(
+                ),
+                partition_by=[F("versions__collection_id")],
+                order_by=F("versions__version_memberships__version_added__pulp_last_updated").desc(
+                    nulls_last=True
+                ),
+            ),
+            repo_version_removed_at=Window(
+                expression=FirstValue(
+                    "versions__version_memberships__version_removed__pulp_last_updated"
+                ),
+                partition_by=[F("versions__collection_id")],
+                order_by=F(
+                    "versions__version_memberships__version_removed__pulp_last_updated"
+                ).desc(nulls_last=True),
+            ),
         )
 
-        highest_versions = defaultdict(
-            lambda: CollectionTuple(None, None, semantic_version.Version("0.0.0"), None)
+        return collections.distinct("versions__collection_id").only(
+            "pulp_created", "name", "namespace"
         )
-        lowest_versions = defaultdict(
-            lambda: CollectionTuple(None, None, semantic_version.Version("100000000000.0.0"), None)
+
+    def paginate_queryset(self, queryset):
+        """Custom pagination."""
+        paginator = self.paginator
+        paginator.count = paginator.get_count(
+            queryset.distinct("versions__collection_id").only("pk")
         )
-        for collection_id, namespace, name, version, pulp_created in versions_qs:
-            version_to_consider = semantic_version.Version(version)
-            collection_tuple = CollectionTuple(namespace, name, version_to_consider, pulp_created)
-            if version_to_consider > highest_versions[collection_id].version:
-                highest_versions[collection_id] = collection_tuple
-            if version_to_consider < lowest_versions[collection_id].version:
-                lowest_versions[collection_id] = collection_tuple
-
-        self.highest_versions_context = highest_versions  # needed by get__serializer_context
-        self.lowest_versions_context = lowest_versions  # needed by get__serializer_context
-        return collections
+        paginator.limit = paginator.get_limit(self.request)
+        if paginator.limit is None:
+            return None
+
+        paginator.offset = paginator.get_offset(self.request)
+        paginator.request = self.request
+        if paginator.count > paginator.limit and paginator.template is not None:
+            paginator.display_page_controls = True
+
+        if paginator.count == 0 or paginator.offset > paginator.count:
+            return []
+        new_queryset = queryset[paginator.offset : paginator.offset + paginator.limit]
+        return list(self.append_context(new_queryset))
 
     def get_object(self):
         """
         Returns a Collection object.
         """
         queryset = self.filter_queryset(self.get_queryset())
+        queryset = self.append_context(queryset)
```

Comment:
I think this and the line above should be combined into one.

Reply:
Actually they can stay as two, but not reassigning the variable would be preferable.